test/lit_tests/diffrules/stablehlo/while3.mlir

```diff
@@ -0,0 +1,98 @@
+// CHECK-NEXT:    "enzyme.set"(%0, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %1 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%1, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %2 = "enzyme.init"() : () -> !enzyme.Cache<tensor<f64>>
+// CHECK-NEXT:    %3 = "enzyme.init"() : () -> !enzyme.Cache<tensor<f64>>
+// CHECK-NEXT:    %4 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%4, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %5 = "enzyme.init"() : () -> !enzyme.Gradient<tensor<f64>>
+// CHECK-NEXT:    "enzyme.set"(%5, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %6:2 = stablehlo.while(%iterArg = %c_1, %iterArg_2 = %arg0) : tensor<i64>, tensor<f64>
+// CHECK-NEXT:     cond {
+// CHECK-NEXT:      %14 = stablehlo.compare  LT, %iterArg, %c : (tensor<i64>, tensor<i64>) -> tensor<i1>
+// CHECK-NEXT:      stablehlo.return %14 : tensor<i1>
+// CHECK-NEXT:    } do {
+// CHECK-NEXT:      "enzyme.push"(%3, %iterArg_2) : (!enzyme.Cache<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      "enzyme.push"(%2, %iterArg_2) : (!enzyme.Cache<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      %14 = stablehlo.multiply %iterArg_2, %iterArg_2 : tensor<f64>
+// CHECK-NEXT:      %15 = stablehlo.add %iterArg, %c_0 : tensor<i64>
+// CHECK-NEXT:      stablehlo.return %15, %14 : tensor<i64>, tensor<f64>
+// CHECK-NEXT:    }
+// CHECK-NEXT:    %7 = "enzyme.get"(%5) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:    %8 = arith.addf %7, %arg1 : tensor<f64>
+// CHECK-NEXT:    "enzyme.set"(%5, %8) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %9 = "enzyme.get"(%5) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:    "enzyme.set"(%5, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:    %10:2 = stablehlo.while(%iterArg = %c_1, %iterArg_2 = %9) : tensor<i64>, tensor<f64>
+// CHECK-NEXT:     cond {
+// CHECK-NEXT:      %14 = stablehlo.compare  LT, %iterArg, %c : (tensor<i64>, tensor<i64>) -> tensor<i1>
+// CHECK-NEXT:      stablehlo.return %14 : tensor<i1>
+// CHECK-NEXT:    } do {
+// CHECK-NEXT:      %14 = stablehlo.add %iterArg, %c_0 : tensor<i64>
+// CHECK-NEXT:      "enzyme.set"(%4, %iterArg_2) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      %15 = "enzyme.get"(%4) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:      "enzyme.set"(%4, %cst) : (!enzyme.Gradient<tensor<f64>>, tensor<f64>) -> ()
+// CHECK-NEXT:      %16 = "enzyme.pop"(%3) : (!enzyme.Cache<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:      %17 = "enzyme.pop"(%2) : (!enzyme.Cache<tensor<f64>>) -> tensor<f64>
+// CHECK-NEXT:      %18 = stablehlo.multiply %15, %17 : tensor<f64>
+// CHECK-NEXT:      %19 = "enzyme.get"(%1) : (!enzyme.Gradient<tensor<f64>>) -> tensor<f64>
```

Comment:
The %1 cache here confuses me. In principle your good design setup such that we don't have any gradient +='s that span the outside of the while regions. Yet here %4 is not able to be mem2reg'd within scope. So something is clearly going awry...